syntax = "proto3";

package dfs;

// The chunk server
service ChunkServer {
  rpc Read (ChunkId) returns (Bytes);
  rpc Write (WriteRequest) returns (Empty);
  rpc Delete (ChunkId) returns (Empty);
  rpc GetChunks (Empty) returns (StringList);
}

// master server
service MasterServer {
  rpc RegisterPeer (RegisterRequest) returns (Empty);
  rpc GetFile (String) returns (ChunkList);
  rpc DeleteFile (String) returns (String);
  rpc GetPeers (Number) returns (StringList);
  rpc AddSegment (SegmentRequest) returns (Empty);
}

// The request message containing block uuid.
message ChunkId {
  string cid = 1;
}

// The request message containing block uuid and data.
message WriteRequest {
  string cid = 1;
  bytes data = 2;
}

message RegisterRequest {
  string ip = 1;
}

message SegmentRequest {
  string sid = 1;
  repeated string chunks = 2;
  repeated string parities = 3;
  repeated string chunk_location = 4;
  repeated string parity_location = 5;
}

message ChunkList {
  map<string, StringList> map = 1;
  repeated string chunks = 2;
  int32 cft = 3;
}
 
message StringList {
  repeated string strs = 1;
}

message Number {
  int32 num = 1;
}

message Bytes {
  bytes data = 1;
}

message String {
  string str = 1;
}

message Bool {
  bool verify = 1;
}

message Empty {
}

